File: src/org/cappuccino_project/ide/intellij/plugin/psi/utils/ObjJExprPsiUtil.java
package org.cappuccino_project.ide.intellij.plugin.psi.utils;

import org.cappuccino_project.ide.intellij.plugin.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ObjJExprPsiUtil {

    public static List<ObjJExpr> allInternalExpressions(@NotNull ObjJExpr expr) {
        List<ObjJExpr> out = new ArrayList<>();
        allInternalExpressions(out, expr);
        return out;
    }

    public static boolean isAssignmentToVariableWithName(@Nullable ObjJExpr expr, @NotNull String variableNameString) {
        return expr != null && expr.getLeftExpr() != null && isLeftExpressionAssignmentToVariable(expr.getLeftExpr(), variableNameString);
    }


    private static boolean isLeftExpressionAssignmentToVariable(@Nullable
                                                                       ObjJLeftExpr leftExpr, @NotNull String variableNameString) {
        if (leftExpr == null) {
            return false;
        }
        if (leftExpr.getVariableAssignmentLogical() != null) {
            final ObjJVariableAssignmentLogical assignment = leftExpr.getVariableAssignmentLogical();
            if (assignment.getQualifiedReference() != null) {
                final String qualifiedName = ObjJVariableNameUtil.getQualifiedNameAsString(assignment.getQualifiedReference(), null);
                return Objects.equals(qualifiedName, variableNameString);
            }
            return assignment.getVariableName() != null && assignment.getVariableName().equals(variableNameString);
        }
        if (leftExpr.getVariableDeclaration() != null) {
            ObjJVariableDeclaration variableDeclaration = leftExpr.getVariableDeclaration();
            for (ObjJQualifiedReference qualifiedReference : variableDeclaration.getQualifiedReferenceList()) {
                final String qualifiedName = ObjJVariableNameUtil.getQualifiedNameAsString(qualifiedReference, null);
                return Objects.equals(qualifiedName, variableNameString);
            }
        }
        return false;
    }


    private static void allInternalExpressions(List<ObjJExpr> out, @Nullable ObjJExpr expr) {
        if (expr == null) {
            return;
        }
        out.add(expr);
        ObjJExpr loopingExpression = expr;
        while (loopingExpression.getExpr() != null) {
            out.add(loopingExpression);
            loopingExpression = loopingExpression.getExpr();
            allInternalExpressions(out, loopingExpression.getLeftExpr());
        }
        for (ObjJExpr loopExpr : new ArrayList<>(out)) {
            for (ObjJRightExpr right : loopExpr.getRightExprList()) {
                allInternalExpressions(out, right);
            }
        }
    }

    private static void allInternalExpressions(List<ObjJExpr> out, @Nullable ObjJLeftExpr leftExpr) {
        if (leftExpr == null) {
            return;
        }
        for (ObjJExpr expr : leftExpr.getExprList()) {
            allInternalExpressions(out, expr);
        }
    }

    private static void allInternalExpressions (List<ObjJExpr> out, @Nullable ObjJRightExpr exprRight) {
        if (exprRight == null) {
            return;
        }
        //allInternalExpressions(out, exprRight.getExpr());
    }

    public static List<ObjJMethodCall> getAllInternalMethodCallExpressions(ObjJExpr expr) {
        List<ObjJExpr> expressions = allInternalExpressions(expr);
        List<ObjJMethodCall> out = new ArrayList<>();
        for (ObjJExpr currentExpression : expressions) {
            getAllInternalMethodCallExpressions(out,currentExpression.getLeftExpr());
        }
        return out;
    }

    private static void getAllInternalMethodCallExpressions(@NotNull List<ObjJMethodCall> out, @Nullable ObjJLeftExpr left) {
        if (left == null) {
            return;
        }
        if (left.getQualifiedReference() != null && left.getQualifiedReference().getMethodCall() != null) {
            out.add(left.getQualifiedReference().getMethodCall());
        }
        if (left.getMethodCall() != null) {
            out.add(left.getMethodCall());
        }
    }
}

File: src/cappuccino/ide/intellij/plugin/psi/utils/ObjJExprPsiUtil.kt
@file:Suppress("UNUSED_PARAMETER")

package cappuccino.ide.intellij.plugin.psi.utils

import cappuccino.ide.intellij.plugin.psi.*

import java.util.ArrayList
import java.util.Objects

fun allInternalExpressions(expr: ObjJExpr): List<ObjJExpr> {
    val out = ArrayList<ObjJExpr>()
    allInternalExpressions(out, expr)
    return out
}

fun isAssignmentToVariableWithName(expr: ObjJExpr?, variableNameString: String): Boolean {
    return expr != null && expr.leftExpr != null && isLeftExpressionAssignmentToVariable(expr.leftExpr, variableNameString)
}


private fun isLeftExpressionAssignmentToVariable(leftExpr: ObjJLeftExpr?, variableNameString: String): Boolean {
    if (leftExpr == null) {
        return false
    }
    if (leftExpr.variableAssignmentLogical != null) {
        val assignment = leftExpr.variableAssignmentLogical
        if (assignment?.qualifiedReference != null) {
            val qualifiedName = ObjJVariableNameUtil.getQualifiedNameAsString(assignment.qualifiedReference, null)
            return qualifiedName == variableNameString
        }
        return assignment?.qualifiedReference?.text == variableNameString
    }
    if (leftExpr.variableDeclaration != null) {
        val variableDeclaration = leftExpr.variableDeclaration
        for (qualifiedReference in variableDeclaration!!.qualifiedReferenceList) {
            val qualifiedName = ObjJVariableNameUtil.getQualifiedNameAsString(qualifiedReference, null)
            return qualifiedName == variableNameString
        }
    }
    return false
}


private fun allInternalExpressions(out: MutableList<ObjJExpr>, expr: ObjJExpr?) {
    if (expr == null) {
        return
    }
    out.add(expr)
    var loopingExpression = expr
    while (loopingExpression?.expr != null) {
        out.add(loopingExpression)
        loopingExpression = loopingExpression.expr
        allInternalExpressions(out, loopingExpression?.leftExpr)
    }
    for (loopExpr in ArrayList(out)) {
        for (right in loopExpr.rightExprList) {
            allInternalExpressions(out, right)
        }
    }
}


private fun allInternalExpressions(out: MutableList<ObjJExpr>, leftExpr: ObjJLeftExpr?) {
    if (leftExpr == null) {
        return
    }
    // @todo figure out how to reimplement this
    /*
    for (expr in leftExpr.) {
        allInternalExpressions(out, expr)
    }*/
}

private fun allInternalExpressions(out: MutableList<ObjJExpr>, exprRight: ObjJRightExpr?) {
    if (exprRight == null) {
        return
    }
    //allInternalExpressions(foldingDescriptors, exprRight.getExpr());
}

fun getAllInternalMethodCallExpressions(expr: ObjJExpr): List<ObjJMethodCall> {
    val expressions = allInternalExpressions(expr)
    val out = ArrayList<ObjJMethodCall>()
    for (currentExpression in expressions) {
        getAllInternalMethodCallExpressions(out, currentExpression.leftExpr)
    }
    return out
}

private fun getAllInternalMethodCallExpressions(out: MutableList<ObjJMethodCall>, left: ObjJLeftExpr?) {
    if (left == null) {
        return
    }
    if (left.qualifiedReference?.methodCall != null) {
        out.add(left.qualifiedReference!!.methodCall!!)
    }
    if (left.methodCall != null) {
        out.add(left.methodCall!!)
    }
}